### PR TITLE
[Storage] Create configuration for encrypting storage

### PR DESCRIPTION
**Pull request by Arminek**
Create a configuration for encrypting storage.

Related with https://github.com/Payum/Payum/issues/634

### CODE REVIEW

**Comment by inssein** [comment]
`storage_encrypting` should probably renamed to `storage_encryption`

File: DependencyInjection/MainConfiguration.php [hunk 1]
```diff
@@ -248,4 +249,22 @@ protected function addDynamicGatewaysSection(ArrayNodeDefinition $dynamicGateway
...
+     * @param ArrayNodeDefinition $rootPrototypeNode
+     */
+    protected function addStorageEncryptingSection(ArrayNodeDefinition $rootPrototypeNode)
+    {
+        $rootPrototypeNode->children()
+            ->arrayNode('storage_encrypting')
+            ->canBeEnabled()
+            ->addDefaultsIfNotSet()
+            ->children()
+                ->booleanNode('enabled')->defaultFalse()->end()
```
Comment (pamil) [inline DependencyInjection/MainConfiguration.php]: Can be skipped as there is `canBeEnabled()` already.

File: DependencyInjection/PayumExtension.php [hunk 2]
```diff
@@ -274,4 +277,29 @@ protected function findSelectedStorageFactoryNameInStorageConfig($storageConfig)
...
+    /**
+     * @param array $config
+     * @param string $id
+     * @param ContainerBuilder $container
+     */
+    protected function decorateStorageWithEncryptingIfConfigured(array $config, $id, ContainerBuilder $container)
+    {
+        if ($config['enabled']) {
+            $container->setParameter('payum.storage.encrypting_algorithm', $config['algorithm']);
+            $container->setParameter('payum.storage.encrypting_initialization_vector', $config['initialization_vector']);
```
Comment (pamil) [inline DependencyInjection/PayumExtension.php]: Missing `secret` parameter to be set.

File: EventListener/CancelGatewayConfigurationUpdateListener.php [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+final class CancelGatewayConfigurationUpdateListener
+{
+    /**
+     * @param LifecycleEventArgs $event
+     */
+    public function preUpdate(LifecycleEventArgs $event)
+    {
+        if ($event->getEntity() instanceof GatewayConfigInterface) {
+            $unityOfWork = $event->getEntityManager()->getUnitOfWork();
```
Comment (pamil) [inline EventListener/CancelGatewayConfigurationUpdateListener.php]: `$unity` -> `$unit`

**Comment by pamil** [comment]
@makasim will you be able to take a look at this one and Payum/Payum#635? Is there anything more we can help with apart from the docs? :)

**Comment by makasim** [comment]
@pamil I will take a look. I'll come back within next few days.

**Comment by pamil** [comment]
@makasim Thanks! :)

File: DependencyInjection/MainConfiguration.php [hunk 4]
```diff
@@ -51,6 +51,7 @@ public function getConfigTreeBuilder()
         ;
         
         $this->addStoragesSection($rootNode);
+        $this->addStorageEncryptingSection($rootNode);
```
Comment (makasim) [inline DependencyInjection/MainConfiguration.php]: that has to go inside `addDynamicGatewaysSection` as we can only encrypt gateway configs for now

https://github.com/Payum/PayumBundle/blob/master/DependencyInjection/MainConfiguration.php#L192

File: EventListener/CancelGatewayConfigurationUpdateListener.php [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+            $unityOfWork->clearEntityChangeSet(spl_object_hash($event->getEntity()));
+        }
```
Comment (makasim) [inline EventListener/CancelGatewayConfigurationUpdateListener.php]: We'd have to think about a better approach. dealing with change sets does not look good imo.

Comment (makasim) [inline EventListener/CancelGatewayConfigurationUpdateListener.php]: I'll think about alternatives

**Comment by makasim** [comment]
closed in favor https://github.com/Payum/PayumBundle/pull/417

**Pull request closed without merging**